Sources/Workspace/UserToolchain.swift

```diff
@@ -365,58 +385,99 @@ public final class UserToolchain: Toolchain {
             }
         }
 
-        // Compute the path of directory containing the PackageDescription libraries.
-        var pdLibDir = ToolchainConfiguration.libDir(forBinDir: binDir)
+        let customSwiftPMLibrariesLocation = try Self.deriveCustomSwiftPMLibrariesLocation(swiftCompilerPath: swiftCompilerPath, destination: destination)
+
+        let xctestPath = Self.deriveXCTestPath()
+
+        self.configuration = .init(
+            swiftCompilerPath: swiftCompilers.manifest,
+            swiftCompilerFlags: self.extraSwiftCFlags,
+            swiftPMLibrariesLocation: customSwiftPMLibrariesLocation,
+            sdkRootPath: self.destination.sdk,
+            xctestPath: xctestPath
+        )
+    }
 
+    private static func deriveCustomSwiftPMLibrariesLocation(
+        swiftCompilerPath: AbsolutePath,
+        destination: Destination
+    ) throws -> ToolchainConfiguration.SwiftPMLibrariesLocation? {
         // Look for an override in the env.
-        if let pdLibDirEnvStr = ProcessEnv.vars["SWIFTPM_PD_LIBS"] {
+        if let pathEnvVariable = ProcessEnv.vars["SWIFTPM_CUSTOM_LIBS_DIR"] ?? ProcessEnv.vars["SWIFTPM_PD_LIBS"] {
+            if ProcessEnv.vars["SWIFTPM_PD_LIBS"] != nil {
+                print("SWIFTPM_PD_LIBS was deprecated in favor of SWIFTPM_CUSTOM_LIBS_DIR")
+            }
             // We pick the first path which exists in an environment variable
             // delimited by the platform specific string separator.
 #if os(Windows)
             let separator: Character = ";"
 #else
             let separator: Character = ":"
 #endif
-            let paths = pdLibDirEnvStr.split(separator: separator).map(String.init)
-            var foundPDLibDir = false
+            let paths = pathEnvVariable.split(separator: separator).map(String.init)
             for pathString in paths {
                 if let path = try? AbsolutePath(validating: pathString), localFileSystem.exists(path) {
-                    pdLibDir = path
-                    foundPDLibDir = true
-                    break
+                    // we found the custom one
+                    return .init(root: path)
                 }
             }
 
-            if !foundPDLibDir {
-                throw InternalError("Couldn't find any SWIFTPM_PD_LIBS directory: \(pdLibDirEnvStr)")
-            }
+            // fail if custom one specified but not found
+            throw InternalError("Couldn't find the custom libraries location defined by SWIFTPM_CUSTOM_LIBS_DIR / SWIFTPM_PD_LIBS: \(pathEnvVariable)")
         }
 
-        var xctestLocation: AbsolutePath?
+        // FIXME: the following logic is pretty fragile, but has always been this way
+        // an alternative cloud be to force explicit locations to always be set explicitly when running in XCode/SwiftPM
+        // debug and assert if not set but we detect that we are in this mode
```

Comment:
Agreed, though I think this code shouldn't be concerned at all with debug mode.  It should know about the default locations, and just expose optional overrides that clients (including `swift-build` etc) can use when they are built in debug mode.

Reply:
But that can be further refined in a later PR.  This is good for now.